### PR TITLE
feat: Signer network interface & in-memory implementation

### LINKED ISSUE #107: [Feature]: Signer network interface
# Feature - Signer network interface

## 1. Description

Define the interface for interacting with the signer network as per the [signer communication LLD](https://docs.google.com/document/d/1q1Ket3ue_H8J3HlvpLhM5-RLRthmw90BugzPwrVpYuw/).

### 2.1 Acceptance Criteria:
   - [ ] The network interface is defined

### PR DESCRIPTION
**Pull request by netrome**
closes #107 and #108 

This PR defines the client interface for the signer network through the `signer::network::Client` trait. It also provides an in-memory implementation of a signer network by wrapping a `tokio::sync::broadcast` channel for illustrative and testing purposes.

### CODE REVIEW

**Review by 8marz8 — APPROVED** [review]

**Comment by 8marz8** [comment]
Just a general question: What will come later as a replacement for the in_memory implementation?

**Comment by netrome** [comment]
> Just a general question: What will come later as a replacement for the in_memory implementation?

Good question, the implementation of the signer network for sBTC-v1 is the HTTP relay https://github.com/stacks-network/sbtc/issues/109 - which is what I'm looking into now.

On a longer horizon I see the possibility of implementing this interface backed by StackerDB. If future requirements change we might take this in another direction and implement a pure peer to peer network based on `libp2p` or a custom implementation. But I don't see that happening any time soon though :)

File: signer/src/network/in_memory.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    async fn receive(&mut self) -> Result<super::Msg, Self::Error> {
+        let mut msg = self.receiver.recv().await?;
+
+        while Some(&msg) == self.recently_sent.get(0) {
+            self.recently_sent.pop_front();
+            msg = self.receiver.recv().await?;
+        }
+
+        Ok(msg)
+    }
```
Comment (djordon) [inline signer/src/network/in_memory.rs:64]: Nice, I see. This is a simple network that works. Everyone receives everyone else's messages and I also receive my own messages. 👍🏽 

An alternative approach, is to give each message a unique ID and filter on IDs that you have broadcasted. That way you do not need to clone the message and you get (roughly) the same guarantees. The approach here is fine too.

File: signer/src/network/in_memory.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+        Client {
+            sender,
+            receiver,
+            recently_sent,
+        }
+    }
+}
+
+impl super::Client for Client {
```
Comment (djordon) [inline signer/src/network/in_memory.rs:47]: I think we should call either the struct or the trait something different than `Client` (I kind of think we should rename them both). Maybe `InternalNetworkClient` for the trait and `MpmcBroadcaster` for the struct. These aren't the best suggestions, I'm just trying to help.

**Review by djordon — COMMENTED** [review]
Looks good, I just have some quibbles with using the names `Client` for the trait and the struct. I'd prefer something a little more informative.

File: signer/src/network/in_memory.rs (hunk 2 above)
Comment (netrome) [reply signer/src/network/in_memory.rs:47]: Sure thing, I'm used to different conventions in different repos and I guess we'll figure out what feels best for us. 

On one hand, I think this works pretty well to reuse simple names such as `Error` or `Client` if the name space is informative. In this scenario, I think `in_memory::Client` looks pretty good in the code. I originally called it `InMemoryClient` which felt a bit repetitive with the namespacing, but works better for people who prefer to import structs by name without any namespacing information. Though I'm personally not a big fan of direct type imports, particularly because it makes it very hard to disambiguate between type imports and bringing traits into scope.

On the other hand, I also think it's ugly to have the same name for the trait and the implementation. Typically traits should be verbs (Clone, Print, Serialize, Encode etc.) - so from this convention it is the trait name that should be updated.

From this reasoning, I would propose renaming the trait to `MessageTransfer`, and keep the in memory network client type name to just `Client`.

Does that sound good to you?

File: signer/src/network/in_memory.rs (hunk 1 above)
Comment (netrome) [reply signer/src/network/in_memory.rs:64]: Yeah, this is a bit of an ugly hack to prevent the signers own messages from echoing back. I like the idea of using IDs instead of matching on the entire message. I guess being able to identify messages will come in handy in the future too. I'll update.

File: signer/src/network/in_memory.rs (hunk 2 above)
Comment (netrome) [reply signer/src/network/in_memory.rs:47]: `MpmcBroadcaster` is more specific than `Client` though. I like that name, so let's use it.

TL; DR
I will rename the trait to `MessageTransfer` and the struct to `MpmcBroadcaster`.

Comment (djordon) [reply signer/src/network/in_memory.rs:47]: > On one hand, I think this works pretty well to reuse simple names such as `Error` or `Client` if the name space is informative. In this scenario, I think `in_memory::Client` looks pretty good in the code. I originally called it `InMemoryClient` which felt a bit repetitive with the namespacing, but works better for people who prefer to import structs by name without any namespacing information. Though I'm personally not a big fan of direct type imports, particularly because it makes it very hard to disambiguate between type imports and bringing traits into scope.

Ahhh, I see. Yeah I can relate, even though I'm definitely guilty of this in my code.

> On the other hand, I also think it's ugly to have the same name for the trait and the implementation. Typically traits should be verbs (Clone, Print, Serialize, Encode etc.) - so from this convention it is the trait name that should be updated.
> 
> From this reasoning, I would propose renaming the trait to `MessageTransfer`, and keep the in memory network client type name to just `Client`.
> 
> Does that sound good to you?

🙏🏽 Yeah that sounds great.

Comment (netrome) [reply signer/src/network/in_memory.rs]: Updated. Let me know what you think.

File: signer/src/network/in_memory.rs (hunk 1 above)
Comment (netrome) [reply signer/src/network/in_memory.rs]: Updated.

**Review by djordon — APPROVED** [review]
Looks good!
